README.rst

```diff
@@ -50,6 +24,102 @@ For python3, use the following command
+
+    from auth0.v3.authentication import GetToken
+
+    token = GetToken('myaccount.auth0.com')
+
+    token.login(client_id='...', client_secret='...', username='user@domain.com', password='secr3t', realm='Username-Password-Authentication')
+
+
+===================
+ID Token validation
+===================
+
+As the result of the authentication and among the credentials received, an ``id_token``
+might be present. This artifact contains information associated to the user that has
+just logged in, provided the scope used contained ``openid``. You can read more
+about ID tokens `here <https://auth0.com/docs/tokens/concepts/id-tokens>`_.
+
+Before you access their contents, you must first verify the ID token to ensure its
+contents has not been tampered with and that is meant for your application to consume.
+
+For that purpose you use the ``TokenVerifier`` class, which requires to be passed
+a few options:
+* A ``SignatureVerifier`` instance, in charge of checking the expected algorithm
+and signature.
+* The expected issuer value, typically matches the Auth0 domain prefixed with
+``https://`` and suffixed with ``/``.
+* The expected audience value, typically matches the Auth0 application client ID.
+
+You choose the signature verifier depending on the signing algorithm used by your Auth0 application.
+You can check its value under ``Advanced settings | OAuth | JsonWebToken Signature Algorithm``.
+* For symmetric algorithms like "HS256", use the `SymmetricSignatureVerifier` class passing
+as secret the client secret value for your Auth0 application.
+* For asymmetric algorithms like "RS256", use the `AsymmetricSignatureVerifier` class passing
+the public URL where the certificates for the public keys can be found.
+
+Auth0 hosts Public Keys inside the ``.well-known`` directory of your tenant's domain.
+That URL looks like this: ``https://myaccount.auth0.com/.well-known/jwks.json``.
+After replacing `myaccount.auth0.com` with your tenant's domain, you should be able
+to access your tenant's public keys.
+
+It is recommended that you make use of asymmetric signing algorithms as their keys are easier
+to rotate in case they need to be revoked.
+
+With all in place, the next snippets shows how to verify an RS256 signed ID token:
+
+.. code-block:: python
+
+    from auth0.v3.authentication.token_verifier import TokenVerifier, AsymmetricSignatureVerifier
+
+    domain = 'myaccount.auth0.com'
+    client_id = 'exampleid'
+
+    # After authenticating
+    id_token = auth_result['id_token']
+
+    jwks_url = 'https://{}/.well-known/jwks.json'.format(domain)
+    issuer = 'https://{}/'.format(domain)
+
+    sv = AsymmetricSignatureVerifier(jwks_url)  # Reusable instance
+    tv = TokenVerifier(signature_verifier=sv, issuer=issuer, audience=client_id)
+    tv.verify(id_token)
+
+Provided something goes wrong, a ``TokenValidationError`` will be raised. In this
+scenario, the ID token should be deemed invalid and its contents not be trusted.
+
```

Comment:
I know this section is just reorganized, but may be worth updating while we're here. Given this is on multiple lines it would be messy to propose changes, consider this, or portions of it:

---

Upon successful authentication, the credentials received may include an ``id_token``, if the authentication request contained the ``openid`` scope. The ``id_token`` contains information associated with the authenticated user. You can read more about ID tokens `here <https://auth0.com/docs/tokens/concepts/id-tokens>`_.

Before you access its contents, you must verify that the ID token has not been tampered with, and that it is meant for your application to consume. The ``TokenVerifier`` class can be used to perform this verification.

To create a ``TokenVerifier``, the following arguments are required:
* A ``SignatureVerifier`` instance, which is responsible for verifying the token's algorithm and signature.
* The expected issuer value, which typically matches the Auth0 domain prefixed with ``https://`` and suffixed with ``/``.
* The expected audience value, which typically matches the Auth0 application client ID.

The type of ``SignatureVerifier`` used depends upon the signing algorithm used by your Auth0 application. You can view this value in your application settings under ``Advanced settings | OAuth | JsonWebToken Signature Algorithm``. Auth0 recommends using the RS256 asymmetric signing algorithm. You can read more about signing algorithms `here <https://auth0.com/docs/tokens/signing-algorithms>`_.

For asymmetric algorithms like RS256, use the ``AsymmetricSignatureVerifier`` class, passing
the public URL where the certificates for the public keys can be found. This will typically be your Auth0 domain with the ``/.well-known/jwks.json`` path appended to it. For example, ``https://your-domain.auth0.com/.well-known/jwks.json``.

For symmetric algorithms like HS256, use the ``SymmetricSignatureVerifier`` class, passing the value of the client secret of your Auth0 application.

The following example demonstrates the verification of an ID token signed with the RS256 signing algorithm:

.. code-block:: python
    
    from auth0.v3.authentication.token_verifier import TokenVerifier, AsymmetricSignatureVerifier
    
    domain = 'myaccount.auth0.com'
    client_id = 'exampleid'
    
    # After authenticating
    id_token = auth_result['id_token']
    
    jwks_url = 'https://{}/.well-known/jwks.json'.format(domain)
    issuer = 'https://{}/'.format(domain)
    
    sv = AsymmetricSignatureVerifier(jwks_url)  # Reusable instance
    tv = TokenVerifier(signature_verifier=sv, issuer=issuer, audience=client_id)
    tv.verify(id_token)
    
If the token verification fails, a ``TokenValidationError`` will be raised. In that scenario, the ID token should be deemed invalid and its contents should not be trusted.

Reply:
Thanks for rewriting this. It looks much better 💯